packages/langium-cli/src/generate.ts

```diff
@@ -53,41 +55,51 @@ function eagerLoad(document: LangiumDocument, uris: Set<string> = new Set()): UR
 }
 
 /**
- * Creates a map that contains all rules of all grammars.
+ * Creates a map that contains elements of all grammars.
  * This includes both input grammars and their transitive dependencies.
  */
-function mapRules(grammars: Grammar[], visited: Set<string> = new Set(), map: Map<Grammar, AbstractRule[]> = new Map()): Map<Grammar, AbstractRule[]> {
+function mapGrammarElements(grammars: Grammar[], visited: Set<string> = new Set(), map: Map<Grammar, GrammarElement[]> = new Map()): Map<Grammar, GrammarElement[]> {
     for (const grammar of grammars) {
         const doc = getDocument(grammar);
         const uriString = doc.uri.toString();
         if (!visited.has(uriString)) {
             visited.add(uriString);
-            map.set(grammar, grammar.rules.map(e => {
-                // Create a new array of rules and copy all rules
-                // Also deactivate all entry rules
-                const shallowCopy = {...e};
-                if (isParserRule(shallowCopy)) {
-                    shallowCopy.entry = false;
-                }
-                return shallowCopy;
-            }));
+            map.set(
+                grammar,
+                (grammar.rules as GrammarElement[])
+                    .concat(grammar.types)
+                    .concat(grammar.interfaces)
+            );
             const importedGrammars = grammar.imports.map(e => resolveImport(documents, e)!);
-            mapRules(importedGrammars, visited, map);
+            mapGrammarElements(importedGrammars, visited, map);
         }
     }
     return map;
 }
 
-function embedReferencedRules(grammar: Grammar, map: Map<Grammar, AbstractRule[]>): void {
+function embedReferencedGrammar(grammar: Grammar, map: Map<Grammar, GrammarElement[]>): void {
     const allGrammars = resolveTransitiveImports(documents, grammar);
     for (const importedGrammar of allGrammars) {
-        const rules = map.get(importedGrammar);
-        if (rules) {
-            grammar.rules.push(...rules);
+        const grammarElements = map.get(importedGrammar) ?? [];
+        for (const element of grammarElements) {
+            const shallowCopy = { ...element };
+            // Deactivate copied entry rule
+            if (isParserRule(shallowCopy)) {
+                shallowCopy.entry = false;
+            }
+            if (isAbstractRule(shallowCopy)) {
+                grammar.rules.push(shallowCopy);
+            } else if (isType(shallowCopy)) {
+                grammar.types.push(shallowCopy);
+            } else if (isInterface(shallowCopy)) {
+                grammar.interfaces.push(shallowCopy);
+            }
```

Comment:
Should we be handling the additional case if something slips through that's not one of these 3? It shouldn't happen, but if it did there's no reporting.